### PR TITLE
Issue 7137: Need a command to get the container Id for a given segment name

### LINKED ISSUE #7137: Need a command to get the container Id for a given segment name
**Is your feature request related to a problem? Please describe.**
In order to get the container Id owning the segment, we need to search in the entire logs to debug any issue.

**Describe the solution you'd like**
Add a new admin cli command, which will give the container Id which is owning the segment.

**Describe alternatives you've considered**
Currently we do not have any cli command to tell the container Id for a given segment name.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by PrajnaU**
**Change log description**  
Adding a new admin cli command, which will give the container Id which is owning the segment.

**Purpose of the change**  
Fixes #7137 

**What the code does**  
Addition of GetContainerIdOfSegmentCommand  is made to Admin-cli to get the container Id of a segment .

**How to verify it**  
Execute the new get-container-id admin command - `segmentstore get-container-id <qualified-segment-name>`
Sample : `segmentstore get-container-id exampleScope/stream2/3.#epoch.0`
Response : `Container Id for the given Segment is :20`

### CODE REVIEW

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            int epoch = Integer.parseInt(segmentArr[2]);
+            long segmentId = NameUtils.computeSegmentId(segmentNumber, epoch);
+            CompletableFuture<WireCommands.StreamSegmentInfo> segmentInfoFuture = segmentHelper.getSegmentInfo(scope, stream, segmentId, super.authHelper.retrieveMasterToken(), 0L);
+            segmentInfoFuture.join();
+            
+            int containerId = segToConMapper.getContainerId(fullyQualifiedSegmentName);
+            output("Container Id for the given Segment is :" + containerId);
+        } catch (Exception ex) {
+
+            output("Error : %s Segment does not exist %s", fullyQualifiedSegmentName, ex);
```
Comment (AnkurAnand11) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:75]: If exception occurs here, does it always mean Segment does not exist ?

Comment (PrajnaU) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:75]: Yes Ankur, as the other conditions for fully qualified segment name and the number of arguments is already checked before. This was added to check if the segment exists or not.

Comment (AnkurAnand11) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:75]: What I meant is, may be there could be connection failure exception in which case this message looks wrong.

Comment (PrajnaU) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java]: Yes, I have updated my PR with the change now.

**Review by AnkurAnand11 — APPROVED** [review]
LGTM

**Review by anishakj — APPROVED** [review]
LGTM

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ *  The arguments for the command that is the fully qualified segment name.
+ */
+public class GetContainerIdOfSegmentCommand extends SegmentStoreCommand {
+
+    private final SegmentToContainerMapper segToConMapper;
+
+    public GetContainerIdOfSegmentCommand(CommandArgs args) {
+        super(args);
+        int containerCount = getServiceConfig().getContainerCount();
+        this.segToConMapper = new SegmentToContainerMapper(containerCount, true);
```
Comment (abhinb) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:44]: does this give the correct containerId if the containercount is wrong for some reason? can you check?

Comment (PrajnaU) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:44]: I checked it now, It is not working if the containercount is wrong.

**Review by abhinb — APPROVED** [review]

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        SegmentHelper segmentHelper = instantiateSegmentHelper(zkClient, pool);
+        final String fullyQualifiedSegmentName = getArg(0);
+        String[] inputParam = fullyQualifiedSegmentName.split("/");
+        Preconditions.checkArgument(inputParam.length == 3, "Invalid qualified-segment-name  '%s'", fullyQualifiedSegmentName);
+        String scope = inputParam[0];
+        String stream = inputParam[1];
+        String[] segmentArr = inputParam[2].split("\\.");
+
+        try {
+            int segmentNumber = Integer.parseInt(segmentArr[0]);
```
Comment (sachin-j-joshi) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:60]: Not sure I completely understand why we need to get StreamSegmentInfo.
why can't we simply call 
`int containerId = segToConMapper.getContainerId(fullyQualifiedSegmentName);` and be done?

Comment (PrajnaU) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:60]: This was done to check if the segment exists or not. As previously even some other values were returned for container id for wrong inputs.

**Review by shshashwat — APPROVED** [review]
LGTM

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java (hunk 3 above)
Comment (sachin-j-joshi) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:60]: Can this check be made optional ? for example accepting another parameter called --skipCheck?
It is not always possible that segment store is healthy or that segment even exists. But fro name we should still be able to get the container ID. This check is completely optional and should not prevent use in such cases

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.curator.framework.CuratorFramework;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Gets the container id of the given segment.
+ * The arguments for the command that is the fully qualified segment name.
+ */
+public class GetContainerIdOfSegmentCommand extends SegmentStoreCommand {
+
+    private final SegmentToContainerMapper segToConMapper;
```
Comment (RaulGracia) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:35]: Nit: Let's use a more understandable naming for this field, such as `segmentToContainerMapper`.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        String[] segmentArr = inputParam[2].split("\\.");
+
+        try {
+            int segmentNumber = Integer.parseInt(segmentArr[0]);
+            int epoch = Integer.parseInt(segmentArr[2]);
+            long segmentId = NameUtils.computeSegmentId(segmentNumber, epoch);
+            CompletableFuture<WireCommands.StreamSegmentInfo> segmentInfoFuture = segmentHelper.getSegmentInfo(scope, stream, segmentId, super.authHelper.retrieveMasterToken(), 0L);
+            segmentInfoFuture.join();
+            int containerId = segToConMapper.getContainerId(fullyQualifiedSegmentName);
+            output("Container Id for the given Segment is :" + containerId);
```
Comment (RaulGracia) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:66]: Nit: ` Segment is :` -> `Segment is: `

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        } catch (Exception ex) {
+            output("Error occurred while fetching containerId : %s", ex);
+              }
```
Comment (RaulGracia) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:69]: Indentation error?

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java (hunk 3 above)
Comment (RaulGracia) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java:60]: Right, whether if a segment exists or not may be optional, as we may want to know the container id of arbitrary segments that may not exist (yet).

Comment (PrajnaU) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java]: I have added skipCheck

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java (hunk 4 above)
Comment (PrajnaU) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java]: I have made the change.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java (hunk 6 above)
Comment (PrajnaU) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java]: I have made the change

**Review by sachin-j-joshi — APPROVED** [review]
LGTM

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                 output("Error occurred while fetching containerId: %s", ex);
+            }
+        }
+    }
+
+    public static CommandDescriptor descriptor() {
+        return new CommandDescriptor(COMPONENT, "get-container-id", "Get the Id of a Container that belongs to a segment.",
+                new ArgDescriptor("qualified-segment-name", "Fully qualified name of the Segment to get info from (e.g., scope/stream/0.#epoch.0)."), new ArgDescriptor("optional:SKIPCHECK", "If given as skipcheck/SKIPCHECK which is optional argument, it skips checking if the segment already exists."));
+    }
+}
```
[the remark this answers is not in the document]
Comment (PrajnaU) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/GetContainerIdOfSegmentCommand.java]: Added empty new line

**Pull request merged**